intro-general.Rmd

```diff
@@ -0,0 +1,107 @@
+A response has a status code indicating whether the request succeeded, headers and it can have a body.
+
+Hopefully the documentation of the web API you are working with shows good examples of responses.
+In any case you'll find yourself experimenting with different requests to see what the response "looks like".
+
+### More resources about HTTP
+
+How do you get started with interacting with HTTP in R?
+
+#### General HTTP resources
+
+* [Mozilla Developer Network docs about HTTP](https://developer.mozilla.org/en-US/docs/Web/HTTP) (recommended in the zine mentioned thereafter)
+* (_not free_) [Julia Evans' Zine "HTTP: Learn your browser's language!"](https://wizardzines.com/zines/http/) 
+* The docs of the web API you are aiming to work with, and a search engine to understand the words that are new.
+
+### HTTP with R
+
+* The docs of the R package you end up choosing!
+* Digging into the source code of another package that does similar things.
+
+
+## HTTP requests in R: what package?
+
+In R, to interact with web resources, it is recommended to use `{curl}`; or its higher-level interfaces `{httr}` ([pronounced _hitter_ or _h-t-t-r_](https://community.rstudio.com/t/pronunciations-of-common-r-terms/1810/15)) or `{crul}`.
+
+Do not use RCurl that is not actively maintained!
+
+When writing a package interacting with web resources, you will probably use either `{httr}` or `{crul}`.
+
+* httr is the most popular and oldest of the two, and supports OAuth among other things. httr docs feature a vignette called [Best practices for API packages](https://httr.r-lib.org/articles/api-packages.html)
+
+* crul does not support OAuth but it uses an object-oriented interface, which you might like.
+crul has a set of [clients, or ways to perform requests](https://docs.ropensci.org/crul/articles/choosing-a-client.html), that might be handy.
+
+Below we will try to programmatically access the [status of GitHub](https://www.githubstatus.com/api/#status), the open-source platfrom provided by the company of the same name.
+We will access the same information with httr and crul.
+If you decide for the low-level curl, feel free to contribute an example.
+
+```{r}
+github_url <- "https://kctbh9vrtdwd.statuspage.io/api/v2/status.json"
+```
+
+The URL above leaves no doubt as to what format the data is provided in, JSON!
+
+Let's first use httr.
+
+```{r}
+response <- httr::GET(github_url)
+
+# Check the response status
+httr::http_status(response)
+
+# Or in a package you'd just write
+httr::stop_for_status(response)
+
+# Parse the content
+httr::content(response)
+
+# In case you wonder, the format was obtained from a header
+httr::headers(response)$`content-type`
+```
+
+Now, the same with crul.
+
+```{r}
+# Create a client and get a response
+client <- crul::HttpClient$new(github_url)
+response <- client$get()
+
+
+# Check the response status
+response$status_http()
+
+# Parse the content
+response$parse()
```

Comment:
@sckott why doesn't "parse()" provide some magic based on HTTP header? E.g. if the header says it's JSON? Is it to limit dependencies?

Reply:
yes, to limit dependencies, to limit complexity, just to leave it to the user

Reply:
👍